core/controllers/reader.py

```diff
@@ -542,33 +542,125 @@ class StateCompleteEventHandler(base.BaseHandler):
+            'state_name': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'session_id': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'time_spent_in_state_secs': {
+                'schema': {
+                    'type': 'float',
+                    'validators': [{
+                        'id': 'is_at_least',
+                        'min_value': 0
+                    }]
+                }
+            }
+        }
+    }
 
     @acl_decorators.can_play_exploration
     def post(self, exploration_id):
         """Handles POST requests."""
-        if self.payload.get('exp_version') is None:
-            raise self.InvalidInputException(
-                'NONE EXP VERSION: State Complete')
         event_services.StateCompleteEventHandler.record(
-            exploration_id, self.payload.get('exp_version'),
-            self.payload.get('state_name'), self.payload.get('session_id'),
-            self.payload.get('time_spent_in_state_secs'))
+            exploration_id, self.normalized_payload.get('exp_version'),
+            self.normalized_payload.get('state_name'),
+            self.normalized_payload.get('session_id'),
+            self.normalized_payload.get('time_spent_in_state_secs'))
         self.render_json({})
 
 
 class LeaveForRefresherExpEventHandler(base.BaseHandler):
     """Tracks a learner leaving an exploration for a refresher exploration."""
 
     REQUIRE_PAYLOAD_CSRF_CHECK = False
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': {
+                'type': 'basestring',
+                'validators': [{
+                    'id': 'is_regex_matched',
+                    'regex_pattern': constants.ENTITY_ID_REGEX
+                }]
+            }
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'exp_version': {
+                'schema': {
+                    'type': 'int',
+                    'validators': [{
+                        'id': 'is_at_least',
+                        # Version must be greater than zero.
+                        'min_value': 1
+                    }]
+                }
+            },
+            'state_name': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'session_id': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'time_spent_in_state_secs': {
+                'schema': {
+                    'type': 'float',
+                    'validators': [{
+                        'id': 'is_at_least',
+                        'min_value': 0
+                    }]
+                }
+            },
+            'refresher_exp_id': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            }
```

Comment:
Should entity ID regex be used for this?

Reply:
@vojtechjelinek Yes updated. PTAL!